<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>


## Introduction


As we discussed in the videos, despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library.

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.


<h2>Regression Models with Keras</h2>

<h3>Objective for this Notebook<h3>    
<h5> 1. How to use the Keras library to build a regression model.</h5>
<h5> 2. Download and Clean dataset </h5>
<h5> 3. Build a Neural Network </h5>
<h5> 4. Train and Test the Network. </h5>     


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1.  <a href="https://#item31">Download and Clean Dataset</a>
2.  <a href="https://#item32">Import Keras</a>
3.  <a href="https://#item33">Build a Neural Network</a>
4.  <a href="https://#item34">Train and Test the Network</a>

</font>
</div>


<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [3]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. 
# If you run this notebook on a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.

#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
#!pip install keras==2.1.6

In [4]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [5]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.


#### Let's check how many data points we have.


In [6]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [7]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [8]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [9]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.


In [10]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [11]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.


In [12]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [13]:
n_cols = predictors_norm.shape[1] # number of predictors

<a id="item1"></a>


<a id='item32'></a>


## Import Keras


Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.


#### Let's go ahead and import the Keras library


In [14]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.


Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.


In [15]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [16]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has one hidden layer, each of 10 hidden units.


<a id="item4"></a>


<a id='item34'></a>


## Train and Test the Network


Let's call the function now to create our model.


In [17]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 50 epochs.


In [18]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 1611.1765 - val_loss: 1157.2321 - 538ms/epoch - 23ms/step
Epoch 2/50
23/23 - 0s - loss: 1572.5214 - val_loss: 1121.7582 - 175ms/epoch - 8ms/step
Epoch 3/50
23/23 - 0s - loss: 1524.9983 - val_loss: 1077.6326 - 100ms/epoch - 4ms/step
Epoch 4/50
23/23 - 0s - loss: 1465.0228 - val_loss: 1022.0052 - 138ms/epoch - 6ms/step
Epoch 5/50
23/23 - 0s - loss: 1388.7983 - val_loss: 954.6179 - 176ms/epoch - 8ms/step
Epoch 6/50
23/23 - 0s - loss: 1293.4381 - val_loss: 874.0405 - 157ms/epoch - 7ms/step
Epoch 7/50
23/23 - 0s - loss: 1180.7203 - val_loss: 781.2676 - 135ms/epoch - 6ms/step
Epoch 8/50
23/23 - 0s - loss: 1050.3998 - val_loss: 684.6041 - 119ms/epoch - 5ms/step
Epoch 9/50
23/23 - 0s - loss: 912.5186 - val_loss: 588.6077 - 155ms/epoch - 7ms/step
Epoch 10/50
23/23 - 0s - loss: 775.0565 - val_loss: 499.8126 - 121ms/epoch - 5ms/step
Epoch 11/50
23/23 - 0s - loss: 647.2549 - val_loss: 423.1394 - 200ms/epoch - 9ms/step
Epoch 12/50
23/23 - 0s - loss: 537.2328 - val_loss

In [28]:
# Increase the number of Epochs to 100 
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Epoch 1/100
23/23 - 0s - loss: 29.9052 - val_loss: 20.6189 - 181ms/epoch - 8ms/step
Epoch 2/100
23/23 - 0s - loss: 29.6964 - val_loss: 22.3341 - 114ms/epoch - 5ms/step
Epoch 3/100
23/23 - 0s - loss: 29.4613 - val_loss: 21.2194 - 137ms/epoch - 6ms/step
Epoch 4/100
23/23 - 0s - loss: 29.3759 - val_loss: 23.5215 - 121ms/epoch - 5ms/step
Epoch 5/100
23/23 - 0s - loss: 29.7036 - val_loss: 22.3100 - 137ms/epoch - 6ms/step
Epoch 6/100
23/23 - 0s - loss: 29.0692 - val_loss: 22.5863 - 137ms/epoch - 6ms/step
Epoch 7/100
23/23 - 0s - loss: 29.1501 - val_loss: 25.0822 - 118ms/epoch - 5ms/step
Epoch 8/100
23/23 - 0s - loss: 28.9862 - val_loss: 23.1073 - 136ms/epoch - 6ms/step
Epoch 9/100
23/23 - 0s - loss: 28.9590 - val_loss: 22.7242 - 156ms/epoch - 7ms/step
Epoch 10/100
23/23 - 0s - loss: 28.7325 - val_loss: 24.9614 - 159ms/epoch - 7ms/step
Epoch 11/100
23/23 - 0s - loss: 28.7971 - val_loss: 25.7300 - 117ms/epoch - 5ms/step
Epoch 12/100
23/23 - 0s - loss: 28.9406 - val_loss: 25.1510 - 119ms/epoch 

Epoch 98/100
23/23 - 0s - loss: 27.9805 - val_loss: 38.2721 - 120ms/epoch - 5ms/step
Epoch 99/100
23/23 - 0s - loss: 27.9107 - val_loss: 34.6608 - 118ms/epoch - 5ms/step
Epoch 100/100
23/23 - 0s - loss: 27.5712 - val_loss: 37.0845 - 133ms/epoch - 6ms/step


In [19]:
# Normalized Dataset 
# Let's import scikit-learn in order to randomly split the data into a training and test sets
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [20]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

In [21]:
# Evaluate Model using test data
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 146.0383


146.03829956054688

In [22]:
#Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

#Let's import the mean_squared_error function from Scikit-learn.
from sklearn.metrics import mean_squared_error

In [23]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

146.0382951156746 0.0


In [24]:
# Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 98.12821960449219
MSE 2: 83.34050750732422
MSE 3: 49.64437484741211
MSE 4: 47.333438873291016
MSE 5: 43.360694885253906
MSE 6: 42.31686019897461
MSE 7: 46.96046447753906
MSE 8: 38.249267578125
MSE 9: 34.98481369018555
MSE 10: 39.07561111450195
MSE 11: 37.33232498168945
MSE 12: 31.475553512573242
MSE 13: 42.6295051574707
MSE 14: 42.0279541015625
MSE 15: 34.08872604370117
MSE 16: 31.10459327697754
MSE 17: 31.592056274414062
MSE 18: 34.15203857421875
MSE 19: 31.72087287902832
MSE 20: 32.5644416809082
MSE 21: 32.39501953125
MSE 22: 33.34033966064453
MSE 23: 29.58626937866211
MSE 24: 33.45945739746094
MSE 25: 36.0087890625
MSE 26: 33.89833450317383
MSE 27: 29.386737823486328
MSE 28: 30.0338191986084
MSE 29: 36.59424591064453
MSE 30: 32.109580993652344
MSE 31: 31.753549575805664
MSE 32: 28.235794067382812
MSE 33: 27.654586791992188
MSE 34: 30.888734817504883
MSE 35: 29.622711181640625
MSE 36: 33.189144134521484
MSE 37: 27.680625915527344
MSE 38: 32.95399856567383
MSE 39: 29.9954242706

In [25]:
# Increase the Number of Hidden Layers:
# Three hidden layers, each of 10 nodes and ReLU activation function
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(2))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Number of hidden layers increased to three.

In [26]:
# Comparison of Errors:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 31.1555


31.155488967895508

In [27]:
# Comparison:
# 1 hidden layer: 10/10 [==============================] - 0s 1ms/step - loss: 146.0383
# 3 hidden layers: 10/10 [==============================] - 0s 1ms/step - loss: 31.1555

146.0383-31.1555

114.88279999999999

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                          |
| ----------------- | ------- | ---------- | ----------------------------------------------------------- |
| 2020-09-21        | 2.0     | Srishti    | Migrated Lab to Markdown and added to course repo in GitLab |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>


This notebook is part of a course on **Coursera** called *Introduction to Deep Learning & Neural Networks with Keras*. If you accessed this notebook outside the course, you can take this course online by clicking [here](https://cocl.us/DL0101EN_Coursera_Week3\_LAB1).


<hr>

Copyright © 2019 [IBM Developer Skills Network](https://cognitiveclass.ai/?utm_medium=dswb&utm_source=bducopyrightlink&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01).
